internal/server/httpapi/trigger.go

```diff
@@ -0,0 +1,413 @@
+package httpapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/gorilla/mux"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/waypoint/internal/clicontext"
+	"github.com/hashicorp/waypoint/internal/serverclient"
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+	"github.com/hashicorp/waypoint/pkg/serverconfig"
+)
+
+// Message is the message we return to the requester when streaming job output
+type Message struct {
+	// The job id that was queued when running the requested trigger
+	JobId string `json:"jobId,omitempty"`
+
+	// Value is the job stream event message to stream back to the requester
+	Value interface{} `json:"value,omitempty"`
+	// ValueType is the kind of job stream event
+	ValueType string `json:"valueType,omitempty"`
+
+	// If the job has completed, we return a 0 for success, 1 for failure.
+	ExitCode string `json:"exitCode,omitempty"`
+	// Error is set when the job failures for any reason
+	Error interface{} `json:"error,omitempty"`
+}
+
+// HandleTrigger will execute a run trigger, if the requested id exists
+// This works by connecting back to our own local gRPC server.
+func HandleTrigger(addr string, tls bool) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		log := hclog.FromContext(ctx)
+		log.SetLevel(hclog.Debug)
+
+		// Authless trigger URLs should be able to make a request
+		// without a token.
+		token := r.URL.Query().Get("token")
+		requireAuth := true
+		if token == "" {
+			log.Trace("no token provided, will attempt to run authless trigger")
+			requireAuth = false
+		}
+
+		// Connect back to our own gRPC service.
+		grpcConn, err := serverclient.Connect(ctx,
+			serverclient.Logger(log),
+			serverclient.FromContextConfig(&clicontext.Config{
+				Server: serverconfig.Client{
+					Address:     addr,
+					RequireAuth: requireAuth,
+					AuthToken:   token,
+
+					// Our gRPC server should always be listening on TLS.
+					// We ignore it because its coming out of our own process.
+					Tls:           tls,
+					TlsSkipVerify: true,
+				},
+			}),
+		)
+		if err != nil {
+			log.Error("trigger connection back to gRPC failed", "err", err)
+			return
+		}
+		defer grpcConn.Close()
+
+		// Our API client
+		client := pb.NewWaypointClient(grpcConn)
+
+		requestVars := mux.Vars(r)
+		runTriggerId := requestVars["id"]
+
+		variablesJSONRaw := r.URL.Query().Get("variables")
+		var (
+			variables         map[string]string
+			variableOverrides []*pb.Variable
+		)
+
+		if variablesJSONRaw != "" {
+			if err := json.Unmarshal([]byte(variablesJSONRaw), &variables); err != nil {
```

Comment:
I would say this feels a little weird to just pass URL-encoded raw JSON as a GET parameter. I'm trying to think if I've seen this done before... in some ways a `variables[key]=value` approach feels like it'd make more sense and something I've seen in other frameworks (Rails for example). Or, base64-encoding the JSON, but that makes it slightly more user hostile.

This could work fine... I'm just thinking...

Reply:
Yep, it's a bit strange. I know people are going to want to override their input vars in some way through this flow, I wasn't sure of a better solution over http 😅 I'm happy to try something better for sure

Reply:
No problem, let's go with this!